book/chapter-01-unix-sockets/01_unix_sockets_intro.md

```diff
@@ -0,0 +1,159 @@
+# UNIX Sockets
+
+## Introduction
+Applications require a way to communicate with eachother: for example, a web app needs to get a list of users from a database, or a messaging app needs to send a message to a friend. This process is known as "Inter-Process Communication" (IPC), and this can be achieved in many ways.
+
```

Comment:
Here you've listed all communication examples where communication can be between computers. Database is usually `TCP` which goes through a network card. Messaging apps will be either `websocket`, `HTTP` etc which will also all go through a network stack. I think maybe it's worth covering the difference between `Inter-Process Communication` that can also support networking, and then highlight the difference. We can then cover  the trade-off is that with unix sockets, piping, and shared memory is that these communication processes are a lot quicker because you don't go through a network stack, and you don't have to bind a port on the localhost. We could then come up with some examples of when we would use unix sockets and shared memory to excite the reader and explain what they will get if they read this section. A good example of unix sockets is that we communicate to the docker daemon via a unix socket. We can build our own daemons for things like logging, or garbage collection. We could also use unix sockets as proxies. For instance, we could have a multithreaded Rust binary (though in this section we won't code it, just explain it), and this Rust binary has a connection pool to a remote database. A python script could send two types of messages to the unix socket for the Rust binary, send, and get. So even though python is single-threaded, it can send requests to the async Rust database proxy, and then get the result later on. The connection via unix socket is then quicker than going through the network stack, and other languages can also make requests to the database proxy daemon including even bashscripts. Interprocess communication enables us to truly seperate processes in an async way, and we can have a lego block system without everything having to be in one language. It also make builds easier because you only have to swap out one component, you don't have to keep rebuilding the proxy database daemon for updates, and it gives teams a clean interface with other processes.

Reply:
Have folded in changes ✅